server/src/main/scala/org/http4s/server/middleware/CSRF.scala

```diff
@@ -49,86 +52,165 @@ import org.http4s.util.{CaseInsensitiveString, encodeHex}
+    RequestCookie(name = cookieName, content = unlift(token))
+
+  /** Extract a `CsrfToken`, if present, from the request,
+    * then try to generate a new token signature, or fail with a validation error
+    * @return newly refreshed token
+    */
+  def refreshedToken[M[_]](r: Request[G])(
+      implicit F: Sync[M]): EitherT[M, CSRFCheckFailed, CSRFToken] =
+    CSRF.cookieFromHeaders(r, cookieName) match {
+      case Some(c) =>
+        EitherT(F.pure(extractRaw(c.content)))
+          .semiflatMap(signToken[M])
+      case None =>
+        EitherT(F.pure(Left(CSRFCheckFailed)))
+    }
+
+  /** Extract a `CsrfToken`, if present, from the request,
+    * then try generate a new token signature, or fail with a validation error.
+    * If not present, generate a new token
+    * @return newly refreshed token
+    */
+  def refreshOrCreate[M[_]](r: Request[G])(
+      implicit F: Sync[M]): EitherT[M, CSRFCheckFailed, CSRFToken] =
+    CSRF.cookieFromHeaders(r, cookieName) match {
+      case Some(c) =>
+        EitherT(F.pure(extractRaw(c.content)))
+          .semiflatMap(signToken[M])
+      case None =>
+        EitherT.liftF(generateToken[M])
+    }
+
+  /** Decode our CSRF token, check the signature
+    * and extract the original token string to sign
     */
-  private[middleware] def extractRaw(token: String): OptionT[F, String] =
-    token.split("-") match {
+  def extractRaw(rawToken: String): Either[CSRFCheckFailed, String] =
+    rawToken.split("-") match {
       case Array(raw, nonce, signed) =>
-        OptionT(F.delay {
-          val mac = Mac.getInstance(CSRF.SigningAlgo)
-          mac.init(key)
-          val out = mac.doFinal((raw + "-" + nonce).getBytes(StandardCharsets.UTF_8))
-          if (MessageDigest.isEqual(out, Base64.getDecoder.decode(signed))) {
-            Some(raw)
-          } else {
-            None
-          }
-        })
+        val mac = Mac.getInstance(CSRF.SigningAlgo)
+        mac.init(key)
+        val out = mac.doFinal((raw + "-" + nonce).getBytes(StandardCharsets.UTF_8))
+        decodeHexString(signed) match {
+          case Some(decoded) =>
+            if (MessageDigest.isEqual(out, decoded)) {
+              Right(raw)
+            } else {
+              Left(CSRFCheckFailed)
+            }
+          case None =>
+            Left(CSRFCheckFailed)
+        }
       case _ =>
-        OptionT.none
+        Left(CSRFCheckFailed)
     }
 
-  /** To be only used on safe methods: if the method is safe (i.e doesn't modify data),
-    * embed a new token if not present, or regenerate the current one to mitigate
-    * BREACH
+  /** To be only used on safe methods: if the method is safe (i.e doesn't modify data)
+    * and a token is present, validate and regenerate it for BREACH to be impractical
     */
-  private[middleware] def validateOrEmbed(r: Request[G], http: Http[F, G]): F[Response[G]] =
+  private[middleware] def validate(r: Request[G], response: F[Response[G]])(
```

Comment:
Odd to take a value in a context, why does this need to be `F[Response[G]]` rather than just a `Response[G]` especially as you just evaluate it inside the comprehension in `flatMap` on both sides of the `match`

Reply:
It can still fail. It's the reason from `F.fromEither(extractRaw(c.content))`